p2p/host/basic/basic_host.go

```diff
@@ -929,6 +931,8 @@ func (h *BasicHost) AllAddrs() []ma.Multiaddr {
 				continue
 			}
 
+			successMapCnt++
+
```

Comment:
Add the port 0 check to line 911 where we check for is unspecifid address. If you do so, it removes the need for introducing this new successMapCnt variable. 

I would prefer keeping this change small because this code is pretty bad and we intend to rewrite it soon in #2229

Reply:
Does the check placed on line 911 mean that the `continue` is executed when the `extMaddr` port is 0? If so, when all `extMaddr` ports are 0, will `observedAddrs` not be added to the `finalAddrs` address list?

Reply:
I don't think so. 
```
		if !manet.IsIPUnspecified(extMaddr) {
				// Add in the mapped addr.
				finalAddrs = append(finalAddrs, extMaddr)
} else {
```
If the nat provided address is not an unspecified addr and(new change) it's not port == 0, we will add it to final addrs. 
We still fall through and add the observed addresses.

Reply:
It is OK to add the address whose port is not 0 to finalAddrs here.

However, at https://github.com/libp2p/go-libp2p/blob/f42120245634b2f6af12f9889859e9f92ac8e588/p2p/host/basic/basic_host.go#L952, it is also necessary to judge whether the port is 0. If the port is 0, should the obsMaddr address be added directly without splicing?